WEB-INTEGRATION.md

```diff
@@ -0,0 +1,678 @@
+  be no causal context, and each `AsyncContext.Variable` would be set to its
+  initial value.
+
+In the cases where the registration web API takes a constructor, the methods and
+getters of the returned object (e.g. the `processor` method in web audio
+worklets, or the `attributeChangedCallback` hook of custom elements) also count
+as callbacks that these APIs invoke, and which should preserve the relevant
+context.
+
+### Stream underlying APIs
+
+The underlying [source](https://streams.spec.whatwg.org/#underlying-source-api),
+[sink](https://streams.spec.whatwg.org/#underlying-sink-api) and
+[transform](https://streams.spec.whatwg.org/#transformer-api) APIs for streams
+are callbacks/methods passed during stream construction. The context in which
+the stream is constructed is then the registration context.
+
+That is also the causal context for the `start` method, but for other methods
+there would be a different causal context, depending on what causes the call to
+that method. For example:
+
+- If `ReadableStreamDefaultReader`’s `read()` method is called, then if the
+  `pull` method is called as a result, then that would be its causal context.
+  This is even if the queue is not empty and the call to `pull` is deferred
+  until previous invocations resolve.
+- If a `Request` is constructed from a `ReadableStream` body, and that is passed
+  to `fetch`, the causal context for the `pull` method invocations should be the
+  context active at the time that `fetch` was called. Similarly, if a response
+  body `ReadableStream` obtained from `fetch` is piped to a `WritableStream`,
+  its `write` method’s causal context is the call to `fetch`.
+
+> TODO: Discuss the details after figuring out events.
+>
+> - `controller.enqueue` inside a `.run()`
+> 
+>    ```js
+>    const readable = new ReadableStream({
+>      pull(controller) {
+>        asyncVar.run(() => {
+>          controller.enqueue(42);
+>        });
+>      }
+>    });
+>    const writable = new WritableStream({
+>      write() {
+>        console.log(asyncVar.get());
+>      }
+>    });
+>    readable.pipeTo(writable);
+>    ```
+>
+> - Transferring streams. Would there be a causal context?
+
+### Observers
+
+Observers are a kind of web API pattern where the constructor for a class takes
+a callback, the instance’s `observe()` method is called to register things that
+should be observed, and then the callback is called when those observations have
+been made.
+
+Unlike FinalizationRegistry, which works similarly, observer callbacks are not
+called once per observation. Instead, multiple observations can be batched into
+one single call. This means that there is not always a single causal context
+that can be used; rather, there might be many.
+
+Given this, for consistency it would be preferable to instead use the
+registration context; that is, the context in which the class is constructed.
+
+- [`MutationObserver`](https://dom.spec.whatwg.org/#mutationobserver)
```

Comment:
As a possible data point, this is already how `console.trace` builds its stack inside MutationObserver, both in Chrome and Firefox.